autoload/ale/autocmd.vim

```diff
@@ -1,77 +1,62 @@
-            \   str2nr(expand('<abuf>')),
-            \   expand('<amatch>')
-            \)
-        endif
-    augroup END
-
-    augroup ALERunOnSaveGroup
-        autocmd!
-        autocmd BufWritePost * call ale#events#SaveEvent(str2nr(expand('<abuf>')))
-    augroup END
+            if g:ale_lint_on_enter
+                autocmd BufWinEnter,BufRead * call ale#Queue(0, 'lint_file', str2nr(expand('<abuf>')))
+                " Track when the file is changed outside of Vim.
+                autocmd FileChangedShellPost * call ale#events#FileChangedEvent(str2nr(expand('<abuf>')))
+            endif
 
-    augroup ALERunOnInsertLeave
-        autocmd!
-        if g:ale_enabled && g:ale_lint_on_insert_leave
-            autocmd InsertLeave * call ale#Queue(0)
-        endif
-    augroup END
+            if g:ale_lint_on_save
+                autocmd BufWritePost * call ale#events#SaveEvent(str2nr(expand('<abuf>')))
+            endif
 
-    augroup ALECursorGroup
-        autocmd!
-        if g:ale_enabled && g:ale_echo_cursor
-            autocmd CursorMoved,CursorHold * call ale#cursor#EchoCursorWarningWithDelay()
-            " Look for a warning to echo as soon as we leave Insert mode.
-            " The script's position variable used when moving the cursor will
-            " not be changed here.
-            autocmd InsertLeave * call ale#cursor#EchoCursorWarning()
-        endif
-    augroup END
+            if g:ale_lint_on_filetype_changed
+                " Only start linting if the FileType actually changes after
+                " opening a buffer. The FileType will fire when buffers are opened.
+                autocmd FileType * call ale#events#FileTypeEvent(str2nr(expand('<abuf>')), expand('<amatch>'))
+            endif
 
-    if !g:ale_enabled
-        augroup! ALERunOnTextChangedGroup
-        augroup! ALERunOnEnterGroup
-        augroup! ALERunOnInsertLeave
-        augroup! ALECursorGroup
+            " Clean up highlights when buffers are hidden.
+            autocmd BufHidden * call ale#highlight#RemoveHighlights()
+            " Clean up buffers automatically when they are unloaded.
+            autocmd BufDelete * call ale#engine#Cleanup(str2nr(expand('<abuf>')))
+            autocmd QuitPre * call ale#events#QuitEvent(str2nr(expand('<abuf>')))
+        augroup END
+
+        augroup ALECursorGroup
+            if g:ale_echo_cursor
+                autocmd CursorMoved,CursorHold * call ale#cursor#EchoCursorWarningWithDelay()
+                " Look for a warning to echo as soon as we leave Insert mode.
+                " The script's position variable used when moving the cursor will
+                " not be changed here.
+                autocmd InsertLeave * call ale#cursor#EchoCursorWarning()
+            endif
+        augroup END
+
+        augroup ALEPatternOptionsGroup
+            autocmd BufEnter,BufRead * call ale#pattern_options#SetOptions(str2nr(expand('<abuf>')))
+        augroup END
```

Comment:
We'll need the pattern options to work even when linting is turned off, for fixing code. That's so you can enable fixing code only for specific buffers, even when linting is off.